src/translate/pddl_parser/parsing_functions.py

```diff
@@ -42,49 +115,92 @@ def set_supertypes(type_list):
         type_name_to_type[desc_name].supertype_names.append(anc_name)
 
 
-def parse_predicate(alist):
-    name = alist[0]
-    arguments = parse_typed_list(alist[1:], only_variables=True)
+def parse_requirements(context, alist):
+    with context.layer("Parsing requirements"):
+        for item in alist:
+            context.check_string(item, item_name="Requirement name")
+        return pddl.Requirements(alist)
+
+
+def construct_type(context: Context, curr_type: str, base_type: str):
+    with context.layer("Parsing PDDL type"):
+        context.check_string(curr_type, item_name="PDDL type")
+        return pddl.Type(curr_type, base_type)
+
+
+def parse_predicate(context, no, alist):
+    syntax = "'(PREDICATE_NAME [VARIABLE [- TYPE]?]*)'"
+    with context.layer(f"Parsing {no}. predicate"):
+        context.check_list(alist, min_size=1, finalmsg=f"Expecting {syntax}.")
+        name = alist[0]
+        context.check_string(name, item_name="Predicate symbol")
+    with context.layer(f"Parsing '{name}' predicate"):
+        arguments = parse_typed_list(context, alist[1:], only_variables=True)
     return pddl.Predicate(name, arguments)
 
 
-def parse_function(alist, type_name):
-    name = alist[0]
-    arguments = parse_typed_list(alist[1:])
+def parse_function(context, alist, type_name):
+    syntax = "'(FUNCTION_NAME [VAR [- TYPE]?]*)'"
+    with context.layer("Parsing function"):
+        context.check_list(alist, min_size=1, finalmsg=f"Expecting {syntax}.")
+        name = alist[0]
+        context.check_string(name, item_name="Function name")
+    with context.layer(f"Parsing function '{name}'"):
+        arguments = parse_typed_list(context, alist[1:])
     return pddl.Function(name, arguments, type_name)
 
 
-def parse_condition(alist, type_dict, predicate_dict):
-    condition = parse_condition_aux(alist, False, type_dict, predicate_dict)
-    return condition.uniquify_variables({}).simplified()
+def parse_condition(context, alist, type_dict, predicate_dict):
+    with context.layer("Parsing condition"):
+        condition = parse_condition_aux(
+            context, alist, False, type_dict, predicate_dict)
+        return condition.uniquify_variables({}).simplified()
 
 
-def parse_condition_aux(alist, negated, type_dict, predicate_dict):
+def parse_condition_aux(context, alist, negated, type_dict, predicate_dict):
     """Parse a PDDL condition. The condition is translated into NNF on the fly."""
+    if len(alist) < 1:
```

Comment:
Add `context.check_list(alist, block_name="CONDITION")` in the line above, if alist is a string here the len check doesn't make sense.